doc/man/66-intree.1.scd
66-intree(1)

# NAME

66-intree - Display information about trees.

# SYNOPSYS

66-intree [ *-h* ] [ *-z* ] [ *-v* _verbosity_ ] [ *-l* _live_ ] [ *-n* ] [ *-o* name,init,enabled,... ] [ *-g* ] [ *-d* _depth_ ] [ *-r* ] _tree_

# DESCRIPTION

*66-intree* displays detailed information about a _tree_.

# OPTIONS

*-h*
	Prints this help.

*-z*
	use color.

*-v* _verbosity_
	Increases/decreases the verbosity of the command.++
	*1* : (Default) Only prints error messages.++
	*2* : Also prints warning messages.++
	*3* : Also prints tracing messages.++
	*4* : Also prints debugging messages.

*-l* _live_
	Changes the supervision directory of _service_ to _live_.
	By default this will be *%%livedir%%*. The default can also be changed at
	compile time by passing the --livedir=_live_ option to *./configure*. An
	existing absolute path is expected and should be within a writable
	filesystem - likely a RAM filesystem. See *66-scandir*(1).

*-n*
	do not display the field name(s) specified.

*-o*
	comma separated list of fields to display. If this option is not passed,
	*66-intree* will display all fields.

*-g*
	shows the dependency list of _tree_ as a graph instead of list.

*-r*
	Shows the dependency list of _tree_ in reverse mode.

*-d* _depth_
	Limits the depth of the dependency list visualisation; defaults to 1.
	This implies the *-g* option.

	If _tree_ is not specified *66-intree* will display information about all
	available trees for the current owner of the process.

## VALID FIELDS FOR -o OPTION

*name*
	Displays the name of the _tree_.

*init*
	Displays a boolean value of the initialization state.

*enabled*
	Displays a boolean value of the enable state.

*start*
	Displays the list of the tree(s) started before.

*current*
	Displays a boolean value of the current state.

*allowed*
	Displays the list of allowed user to use the _tree_.

*symlinks*
	Displays the target of the _tree_'s symlinks.

*contents*
	Displays the contents of the _tree_.

# COMMAND AND OUTPUT EXAMPLE

The command "66-intree boot", run as root user, on _Obarun_'s default system
displays the following where *boot* is the tree used to properly boot 
the machine:

```
Name         : boot
Initialized  : yes
Enabled      : no
Starts after : None
Allowed      : root
Symlinks     : svc->source db->source
Current      : no
Contains     : tty12  system-hostname  mount-run  populate-run  mount-tmp
               populate-tmp  mount-proc  mount-sys  populate-sys  mount-dev
               mount-pts  mount-shm  populate-dev  mount-cgroups  00
               modules-kernel  udevd  udevadm  devices-crypttab
               system-hwclock  system-random  modules-system  system-sysctl
               system-fontnkey  devices-dmraid  devices-btrfs  devices-lvm
               devices-zfs  system-Devices  mount-swap  all-Mount
               system-fsck  mount-fstab  all-System  mount-rw  local-iptables
               local-ip6tables  local-loop  local-sethostname  local-time
               local-authfiles  local-tmpfiles  local-rc  local-dmesg  all-Local
               all-Runtime  All
```

The field *name* gives you the name of the _tree_.

The field *Initialized* tells you if the _tree_ was initialized with "66-init" tool.

The field *Enabled* reveals the state of the tree -- see "66-tree -E".

The field *Starts after* reveals the start process order if the _tree_ is enabled,
meaning which _tree_ is started before the current one.

The field *Current* tells you if the _tree_ is the current one or not -- see
"66-tree -c".

The field *Allowed* gives you a list of user(s) allowed to handle the 
_tree_ -- see "66-tree -[a|b]".

The field *Symlinks* tells you if the current live state point to the 
*source* or the *backup* of the _tree_. Every use of  "66-enable" tool 
create a automatic *backup* of the _tree_ for *classic*,*bundle* or *atomic* 
service(s). A symlink pointing to *backup* mean that you have enabled a service
without starting it. Right after a boot, each tree should point to *source*.

The field *Contains* gives you a list of all services enabled in the _tree_.

You can display the contents list as a graph and only these fields using 
the command "66-intree -o contents -g boot":

```
Contents    : /
              ├─(253,Enabled,classic) tty12
              ├─(0,Enabled,oneshot) system-hostname
              ├─(0,Enabled,oneshot) mount-run
              ├─(0,Enabled,oneshot) populate-run
              ├─(0,Enabled,oneshot) mount-tmp
              ├─(0,Enabled,oneshot) populate-tmp
              ├─(0,Enabled,oneshot) mount-proc
              ├─(0,Enabled,oneshot) mount-sys
              ├─(0,Enabled,oneshot) populate-sys
              ├─(0,Enabled,oneshot) mount-dev
              ├─(0,Enabled,oneshot) mount-pts
              ├─(0,Enabled,oneshot) mount-shm
              ├─(0,Enabled,oneshot) populate-dev
              ├─(0,Enabled,oneshot) mount-cgroups
              ├─(0,Enabled,bundle) 00
              ├─(0,Enabled,oneshot) modules-kernel
              ├─(485,Enabled,longrun) udevd
              ├─(0,Enabled,oneshot) udevadm
              ├─(0,Enabled,oneshot) devices-crypttab
              ├─(0,Enabled,oneshot) system-hwclock
              ├─(0,Enabled,oneshot) system-random
              ├─(0,Enabled,oneshot) modules-system
              ├─(0,Enabled,oneshot) system-sysctl
              ├─(0,Enabled,oneshot) system-fontnkey
              ├─(0,Enabled,oneshot) devices-dmraid
              ├─(0,Enabled,oneshot) devices-btrfs
              ├─(0,Enabled,oneshot) devices-lvm 
              ├─(0,Enabled,oneshot) devices-zfs
              ├─(0,Enabled,bundle) system-Devices
              ├─(0,Enabled,oneshot) mount-swap 
              ├─(0,Enabled,bundle) all-Mount
              ├─(0,Enabled,oneshot) system-fsck
              ├─(0,Enabled,oneshot) mount-fstab
              ├─(0,Enabled,bundle) all-System
              ├─(0,Enabled,oneshot) mount-rw
              ├─(0,Enabled,oneshot) local-iptables
              ├─(0,Enabled,oneshot) local-ip6tables
              ├─(0,Enabled,oneshot) local-loop
              ├─(0,Enabled,oneshot) local-sethostname
              ├─(0,Enabled,oneshot) local-time
              ├─(0,Enabled,oneshot) local-authfiles
              ├─(0,Enabled,oneshot) local-tmpfiles  
              ├─(0,Enabled,oneshot) local-rc
              ├─(0,Enabled,oneshot) local-dmesg
              ├─(0,Enabled,bundle) all-Local
              ├─(0,Enabled,oneshot) all-Runtime
              └─(0,Enabled,bundle) All

```
For each service the first field found between *()* parentheses is the 
corresponding pid of the service, the second one is the state of the 
service, and next to it is the type of the service, separated by commas,
and finally the name of the service is displayed after the parenthesis ')'.

By default the dependency graph is rendered in the order of execution. In 
this example the *classic tty12* is the first executed service and 
*bundle All* is the last before it finishes. You can reverse the rendered order with
the *-r* option.

You can display the name and current field and only these fields for each 
tree using the command "66-intree -o name,current":

```
Name        : boot
Current     : no

Name        : docker
Current     : no

Name        : root
Current     : no

Name        : test
Current     : yes

Name        : user
Current     : no

```


NEWS
Changelog for 66

In 0.2.5.2
----------

    - Fix build: Remove bytes.h oblibs header file
    
In 0.2.5.1
----------
    
    - Bugs fix
    - 66-tree -S options: if after_tree and tree have the same name,
    tree is considered as the very first tree to start.
    - 66-unmountall: do not umount SS_LIVE
    - Add SIGPWR control file at creation of .s6-svscan directory

In 0.2.5.0
----------
    
    - Bugs Fix: Bad memory allocation
    - 66-tree: add -S option which allow to order the start tree process
    even after it creation.
    
In 0.2.4.1
----------
    
    - Hot fix: fix html documentation

In 0.2.4.0
----------
    
    - Bugs fix: Bad memory allocation.
    
    - Enable again RB_DISABLE_CAD but don't crash if its not available
    and warn user
    
    - Add @optsdepends and @extdepends field at [main] section:
        - @optsdepends can be considere as: "enable one on this service 
        or none"
        - @extdepends can be considere as: "enable the service if it is 
        not already declared on a tree"
    
    - 66-in{tree,service}: 
        - in case of empty value the tools return None
        - add -n option: this avoids to display the name of the field.
        - add field at 66-intree:
            - start: displays the list of tree(s) started before
            - allowed: displays a list of allowed user to use the tree
            - symlinks: displays the target of tree's symlinks
        - add field at 66-inservice:
            - optsdepends: displays the service's optional dependencies
            - extdepends: displays the service's external dependencies
    
    - 66-shutdown skeleton: be safier and check if options are past
    
    - New tool:
        - 66-update: The 66-update program makes a complete transition 
        of trees and the live directory using a old 66 format 
        (the one being replaced) with the new 66 format.

In 0.2.3.2
----------
    
    - Fix bad annoucement at 66-svctl
    - Warn in case of bad key at parenthesis parsing process
    
In 0.2.3.1
----------
    
    - Bugs fix: bad allocation memory
    - Makefile: remove creation of empty directories. 66-tree will check
    all directories and create it if it missing.
    
In 0.2.3.0
----------
    
    - Adapt to oblibs 0.0.4.0
    - Remove RB_DISABLE_CAD at 66-boot by default and allow to enable
    it with -c option. Starting on virtual system like lxc crash with
    this option. Well be safe on every case.
    - Use of new function log_? familly from oblibs:
        - standardization of the exit code
        - standardization of the verbosity output
        - allow verbosity to 4 to display debug message
    - Doc typo fix        
    
In 0.2.2.2
----------
    
    - Adapt to skalibs 2.9.1.0
    - Remove insta_? deprecated function
    
In 0.2.2.1
----------
    
    - Typo fix at html documentation
    - Fix 66-intree without options
    
In 0.2.2.0
----------

    - Bugs fix: Always check the existing of the 66 heart directories
    - 66-info is now deprecated and splitted into two differents API.
    66-intree give informations about tree
    66-inservice give informations about services

In 0.2.1.2
----------

    - Hot fix: fix 66-info graph display.
    
In 0.2.1.1
----------

    - Hot fix: fix the build of the service dependencies graph.

In 0.2.1.0
----------

    - Bugs fix.
   
    - A synchronization is now made on reboot even with force option.
   
    - The field @name has no effects except for instantiated service but 
    can be omitted. The name of the frontend file is took as name. 
    In case of instantiated service, the field @name must contain the 
    complete name of the frontend service file -- Refer to the frontend 
    documentation page.
    
    - New skeleton file: rc.shutdown.final. This skeleton file will be run
    at the very end of the shutdown procedure, after all processes have 

src/66/66-intree.c
/* 
 * 66-intree.c
 * 
 * Copyright (c) 2018-2020 Eric Vidal <eric@obarun.org>
 * 
 * All rights reserved.
 * 
 * This file is part of Obarun. It is subject to the license terms in
 * the LICENSE file found in the top-level directory of this
 * distribution.
 * This file may not be copied, modified, propagated, or distributed
 * except according to the terms contained in the LICENSE file./
 */

#include <string.h>
#include <locale.h>
#include <langinfo.h>
#include <sys/types.h>
#include <wchar.h>
#include <unistd.h>//access

#include <oblibs/sastr.h>
#include <oblibs/log.h>
#include <oblibs/obgetopt.h>
#include <oblibs/types.h>
#include <oblibs/string.h>
#include <oblibs/files.h>

#include <skalibs/stralloc.h>
#include <skalibs/genalloc.h>
#include <skalibs/lolstdio.h>
#include <skalibs/bytestr.h>
#include <skalibs/buffer.h>

#include <66/info.h>
#include <66/utils.h>
#include <66/constants.h>
#include <66/tree.h>
#include <66/enum.h>
#include <66/resolve.h>
#include <66/backup.h>

static unsigned int REVERSE = 0 ;
static unsigned int NOFIELD = 1 ;
static unsigned int GRAPH = 0 ;
static uid_t OWNER ;
static char OWNERSTR[UID_FMT] ;

static stralloc base = STRALLOC_ZERO ;
static stralloc live = STRALLOC_ZERO ;
static stralloc src = STRALLOC_ZERO ;

static wchar_t const field_suffix[] = L" :" ;
static char fields[INFO_NKEY][INFO_FIELD_MAXLEN] = {{ 0 }} ;
static void info_display_name(char const *field,char const *treename) ;
static void info_display_init(char const *field,char const *treename) ;
static void info_display_order(char const *field,char const *treename) ;
static void info_display_enabled(char const *field,char const *treename) ;
static void info_display_current(char const *field,char const *treename) ;
static void info_display_allow(char const *field,char const *treename) ;
static void info_display_symlink(char const *field,char const *treename) ;
static void info_display_contents(char const *field,char const *treename) ;
ss_resolve_graph_style *STYLE = &graph_default ;

info_opts_map_t const opts_tree_table[] =
{
	{ .str = "name", .func = &info_display_name, .id = 0 },
	{ .str = "init", .func = &info_display_init, .id = 1 },
	{ .str = "enabled", .func = &info_display_enabled, .id = 2 },
	{ .str = "start", .func = &info_display_order, .id = 3 },
	{ .str = "current", .func = &info_display_current, .id = 4 },
	{ .str = "allowed", .func = &info_display_allow, .id = 5 },
	{ .str = "symlinks", .func = &info_display_symlink, .id = 6 },
	{ .str = "contents", .func = &info_display_contents, .id = 7},
	{ .str = 0, .func = 0, .id = -1 }
} ;

#define MAXOPTS 9
#define checkopts(n) if (n >= MAXOPTS) log_die(100, "too many options")
#define DELIM ','

#define USAGE "66-intree [ -h ] [ -z ] [ -v verbosity ] [ -l live ] [ -n ] [ -o name,init,enabled,... ] [ -g ] [ -d depth ] [ -r ] tree"

static inline void info_help (void)
{
  static char const *help =
"66-intree <options> tree \n"
"\n"
"options :\n"
"	-h: print this help\n"
"	-z: use color\n"
"	-v: increase/decrease verbosity\n"
"	-l: live directory\n"
"	-n: do not display the names of fields\n"
"	-o: comma separated list of field to display\n"
"	-g: displays the contents field as graph\n"
"	-d: limit the depth of the contents field recursion\n"
"	-r: reverse the contents field\n" 
"\n"
"valid field for -o options are:\n"
"\n"
"	name: displays the name of the tree\n"
"	init: displays a boolean value of the initialization state\n"
"	enabled: displays a boolean value of the enable state\n"
"	start: displays the list of tree(s) started before\n"
"	current: displays a boolean value of the current state\n"
"	allowed: displays a list of allowed user to use the tree\n"
"	symlinks: displays the target of tree's symlinks\n"
"	contents: displays the contents of the tree\n"
"\n"
;

 if (buffer_putsflush(buffer_1, help) < 0)
    log_dieusys(LOG_EXIT_SYS, "write to stdout") ;
}

static int info_cmpnsort(stralloc *sa)
{
	size_t pos = 0 ;
	if (!sa->len) return 0 ;
	size_t salen = sa->len ;
	size_t nel = sastr_len(sa), idx = 0, a = 0, b = 0 ;
	char names[nel][4096] ;
	char tmp[4096] ;
	
	for (; pos < salen && idx < nel ; pos += strlen(sa->s + pos) + 1,idx++)
	{
		memcpy(names[idx],sa->s + pos,strlen(sa->s + pos)) ;
		names[idx][strlen(sa->s+pos)] = 0 ;
	}
	for (; a < nel - 1 ; a++)
	{
		for (b = a + 1 ; b < idx ; b++)
		{
			if (strcmp(names[a],names[b]) > 0)
			{
				strcpy(tmp,names[a]) ;
				strcpy(names[a],names[b]);
				strcpy(names[b],tmp);
			}
		}
	}
	sa->len = 0 ;
	for (a = 0 ; a < nel ; a++)
	{
		if (!sastr_add_string(sa,names[a])) return 0 ;
	}	
	return 1 ;
}

static void info_display_name(char const *field, char const *treename)
{
	if (NOFIELD) info_display_field_name(field) ;
	if (!bprintf(buffer_1,"%s",treename)) 
		log_dieusys(LOG_EXIT_SYS,"write to stdout") ;
	if (buffer_putsflush(buffer_1,"\n") == -1) 
		log_dieusys(LOG_EXIT_SYS,"write to stdout") ;
}

static void info_display_init(char const *field,char const *treename)
{
	unsigned int init = 0 ;
	int r = set_livedir(&live) ;
	if (!r) log_die_nomem("stralloc") ;
	if (r == -1) log_die(LOG_EXIT_SYS,"live: ",live.s," must be an absolute path") ;
	
	if (!stralloc_cats(&live,SS_STATE + 1) ||
	!stralloc_cats(&live,"/") ||
	!stralloc_cats(&live,OWNERSTR) ||
	!stralloc_cats(&live,"/") ||
	!stralloc_cats(&live,treename) ||
	!stralloc_cats(&live,"/init") ||
	!stralloc_0(&live)) log_die_nomem("stralloc") ;
	if (!access(live.s, F_OK)) init = 1 ;

	if (NOFIELD) info_display_field_name(field) ;
	if (!bprintf(buffer_1,"%s%s%s",init ? log_color->valid : log_color->warning, init ? "yes":"no",log_color->off)) 
		log_dieusys(LOG_EXIT_SYS,"write to stdout") ;
	
	if (buffer_putsflush(buffer_1,"\n") == -1)
		log_dieusys(LOG_EXIT_SYS,"write to stdout") ;

}

static void info_display_current(char const *field,char const *treename)
{
	stralloc sacurr = STRALLOC_ZERO ;
	int current = 0 ;
	
	if (tree_find_current(&sacurr,base.s,OWNER))
	{
		char name[sacurr.len + 1] ;//be paranoid +1
		if (!ob_basename(name,sacurr.s)) log_dieu(LOG_EXIT_SYS,"basename of: ",sacurr.s) ;
		current = obstr_equal(treename,name) ;
	}
	if (NOFIELD) info_display_field_name(field) ;
	if (!bprintf(buffer_1,"%s%s%s", current ? log_color->blink : log_color->warning, current ? "yes":"no",log_color->off))
		log_dieusys(LOG_EXIT_SYS,"write to stdout") ;
	
	if (buffer_putsflush(buffer_1,"\n") == -1) 
		log_dieusys(LOG_EXIT_SYS,"write to stdout") ;	
	
	stralloc_free(&sacurr) ;
}

static void info_display_enabled(char const *field,char const *treename)
{
	int enabled = tree_cmd_state(VERBOSITY,"-s",treename) ;
	if (NOFIELD) info_display_field_name(field) ;
	if (!bprintf(buffer_1,"%s%s%s",enabled == 1 ? log_color->valid : log_color->warning, enabled == 1 ? "yes":"no",log_color->off)) 
		log_dieusys(LOG_EXIT_SYS,"write to stdout") ;
	
	if (buffer_putsflush(buffer_1,"\n") == -1) 
		log_dieusys(LOG_EXIT_SYS,"write to stdout") ;
}

static void info_display_order(char const *field,char const *treename)
{
	int r ;
	size_t pos = 0 ;
	stralloc contents = STRALLOC_ZERO ;
	stralloc tmp = STRALLOC_ZERO ;
	if (NOFIELD) info_display_field_name(field) ;
	int enabled = tree_cmd_state(VERBOSITY,"-s",treename) ;
	
	r = file_readputsa(&tmp,src.s,SS_STATE + 1) ;
	if(!r) log_dieusys(LOG_EXIT_SYS,"open: ", src.s,SS_STATE) ;
	
	if (tmp.len && enabled == 1)
	{
		if (!sastr_rebuild_in_oneline(&tmp) ||
		!sastr_clean_element(&tmp))
			log_dieusys(LOG_EXIT_SYS,"rebuilt state list") ;
		for (pos = 0 ;pos < tmp.len; pos += strlen(tmp.s + pos) + 1)
		{
			char *name = tmp.s + pos ;
			if (obstr_equal(name,treename)) break ;
			if (!sastr_add_string(&contents,name))
				log_dieusys(LOG_EXIT_SYS,"rebuilt state list") ;
		}
	}
	if (contents.len) {
		 info_display_list(field,&contents) ;
	}
	else
	{
		if (!bprintf(buffer_1,"%s%s%s",log_color->warning,"None",log_color->off)) 
			log_dieusys(LOG_EXIT_SYS,"write to stdout") ;
	
		if (buffer_putsflush(buffer_1,"\n") == -1) 
			log_dieusys(LOG_EXIT_SYS,"write to stdout") ;
	}
	stralloc_free(&contents) ;
	stralloc_free(&tmp) ;
}

static void info_get_graph_src(ss_resolve_graph_t *graph,char const *src,unsigned int reverse)
{
	
	size_t srclen = strlen(src), pos ;
	stralloc sa = STRALLOC_ZERO ;
	ss_resolve_t res = RESOLVE_ZERO ;
	char solve[srclen + SS_RESOLVE_LEN + 1] ;
	memcpy(solve,src,srclen) ;
	memcpy(solve + srclen, SS_RESOLVE,SS_RESOLVE_LEN) ;
	solve[srclen + SS_RESOLVE_LEN] = 0 ;
	
	if (!sastr_dir_get(&sa,solve,"",S_IFREG))
		log_dieusys(LOG_EXIT_SYS,"get source service file at: ",solve) ;
	
	for (pos = 0 ;pos < sa.len; pos += strlen(sa.s + pos) + 1)
	{
		char *name = sa.s + pos ;
		if (!ss_resolve_read(&res,src,name))
			log_dieu(LOG_EXIT_SYS,"read resolve file of: ",name) ;
		if (!ss_resolve_graph_build(graph,&res,src,reverse)) 
			log_dieu(LOG_EXIT_SYS,"build the graph from: ",src) ; 
	}
	
	stralloc_free(&sa) ;
	ss_resolve_free(&res) ;
}

static void info_display_allow(char const *field, char const *treename)
{
	if (NOFIELD) info_display_field_name(field) ;
	size_t treenamelen = strlen(treename) , pos ;
	stralloc sa = STRALLOC_ZERO ;
	stralloc salist = STRALLOC_ZERO ;
	char tmp[src.len + treenamelen + SS_RULES_LEN + 1 ] ;
	// force to close the string
	auto_strings(tmp,src.s) ;
	auto_string_from(tmp,src.len,treename,SS_RULES) ;

	if (!sastr_dir_get(&sa,tmp,"",S_IFREG))
		log_dieusys(LOG_EXIT_SYS,"get permissions of tree at: ",tmp) ;
	
	for (pos = 0 ;pos < sa.len; pos += strlen(sa.s + pos) + 1)
	{
		char *suid = sa.s + pos ;
		uid_t uid = 0 ;
		if (!uid0_scan(suid, &uid))
			log_dieusys(LOG_EXIT_SYS,"get uid of: ",suid) ;
		if (pos) 
			if (!stralloc_cats(&salist," ")) log_die_nomem("stralloc") ;
		if (!get_namebyuid(uid,&salist))
			log_dieusys(LOG_EXIT_SYS,"get name of uid: ",suid) ;
	}
	if (!stralloc_0(&salist)) log_die_nomem("stralloc") ;
	if (!sastr_rebuild_in_oneline(&salist)) log_dieu(LOG_EXIT_SYS,"rebuild list: ",salist.s) ;
	if (!stralloc_0(&salist)) log_die_nomem("stralloc") ;
	info_display_list(field,&salist) ;
	
	stralloc_free(&sa) ;
	stralloc_free(&salist) ;
}

static void info_display_symlink(char const *field, char const *treename)
{
	if (NOFIELD) info_display_field_name(field) ;
	ssexec_t info = SSEXEC_ZERO ;
	if (!auto_stra(&info.treename,treename)) log_die_nomem("stralloc") ;
	if (!auto_stra(&info.base,base.s)) log_die_nomem("stralloc") ;
	int db , svc ;
	size_t typelen ;
	char type[UINT_FMT] ;
	typelen = uint_fmt(type, TYPE_BUNDLE) ;
	type[typelen] = 0 ;
	
	char cmd[typelen + 6 + 1] ;
	
	auto_strings(cmd,"-t",type," -b") ;
	db = backup_cmd_switcher(VERBOSITY,cmd,&info) ;
	if (db < 0) log_dieu(LOG_EXIT_SYS,"find realpath of symlink for db of tree: ",info.treename.s) ;
	
	typelen = uint_fmt(type, TYPE_CLASSIC) ;
	type[typelen] = 0 ;
		
	auto_strings(cmd,"-t",type," -b") ;
	svc = backup_cmd_switcher(VERBOSITY,cmd,&info) ;
	if (svc < 0) log_dieu(LOG_EXIT_SYS,"find realpath of symlink for svc of tree: ",info.treename.s) ;

	if (!bprintf(buffer_1,"%s%s%s%s%s%s%s%s", "svc->",!svc ? log_color->valid : log_color->warning , !svc ? "source" : "backup",log_color->off, " db->", !db ? log_color->valid : log_color->warning, !db ? "source" : "backup", log_color->off)) 
		log_dieusys(LOG_EXIT_SYS,"write to stdout") ;
	
	if (buffer_putsflush(buffer_1,"\n") == -1) 
		log_dieusys(LOG_EXIT_SYS,"write to stdout") ;
	
	ssexec_free(&info) ;	
}

static void info_display_contents(char const *field, char const *treename)
{
	int r ;
	size_t padding = 1, treenamelen = strlen(treename) ;
	ss_resolve_t res = RESOLVE_ZERO ;
	ss_resolve_graph_t graph = RESOLVE_GRAPH_ZERO ;
	stralloc salist = STRALLOC_ZERO ;
	
	char tmp[src.len + treenamelen + SS_SVDIRS_LEN + 1] ;
	memcpy(tmp,src.s,src.len) ;
	memcpy(tmp + src.len,treename,treenamelen) ;
	memcpy(tmp + src.len + treenamelen,SS_SVDIRS,SS_SVDIRS_LEN) ;
	tmp[src.len + treenamelen + SS_SVDIRS_LEN] = 0 ;
	
	info_get_graph_src(&graph,tmp,0) ;
	
	if (NOFIELD) padding = info_display_field_name(field) ;
	else { field = 0 ; padding = 0 ; }
	
	if (!genalloc_len(ss_resolve_t,&graph.name)) goto empty ;
	
	r = ss_resolve_graph_publish(&graph,0) ;
	if (r < 0) log_die(LOG_EXIT_USER,"cyclic graph detected at tree: ", treename) ;
	else if (!r) log_dieusys(LOG_EXIT_SYS,"publish service graph of tree: ",treename) ;
	
	for (size_t i = 0 ; i < genalloc_len(ss_resolve_t,&graph.sorted) ; i++)
	{
		char *string = genalloc_s(ss_resolve_t,&graph.sorted)[i].sa.s ;
		char *name = string + genalloc_s(ss_resolve_t,&graph.sorted)[i].name ;
		if (!stralloc_catb(&salist,name,strlen(name)+1)) log_die_nomem("stralloc") ;
	}
		
	if (GRAPH)
	{
		if (!bprintf(buffer_1,"%s\n","/")) 
			log_dieusys(LOG_EXIT_SYS,"write to stdout") ;
		size_t el = sastr_len(&salist) ;
		if (!sastr_rebuild_in_oneline(&salist)) log_dieu(LOG_EXIT_SYS,"rebuild dependencies list") ;
		ss_resolve_init(&res) ;
		res.ndeps = el ;
		res.deps = ss_resolve_add_string(&res,salist.s) ;
		if (!info_graph_init(&res,tmp,REVERSE, padding, STYLE))
			log_die(LOG_EXIT_SYS,"display graph of: ",treename) ;
		goto freed ;
	}
	else
	{
		if (REVERSE) 
			if (!sastr_reverse(&salist))
				log_dieusys(LOG_EXIT_SYS,"reverse dependencies list") ;
		info_display_list(field,&salist) ;
		goto freed ;
	}
	empty:
		if (GRAPH)
		{
			if (!bprintf(buffer_1,"%s\n","/")) 
				log_dieusys(LOG_EXIT_SYS,"write to stdout") ;
			if (!bprintf(buffer_1,"%*s%s%s",padding,"",STYLE->last,"None")) 
				log_dieusys(LOG_EXIT_SYS,"write to stdout") ;
		}
		else
		{
			if (!bprintf(buffer_1,"%s","None")) 
				log_dieusys(LOG_EXIT_SYS,"write to stdout") ;
		}
		if (buffer_putsflush(buffer_1,"\n") == -1)
			log_dieusys(LOG_EXIT_SYS,"write to stdout") ;
	freed:
		ss_resolve_free(&res) ;
		ss_resolve_graph_free(&graph) ;
		stralloc_free(&salist) ;
}

static void info_display_all(char const *treename,int *what)
{
	
	unsigned int i = 0 ;
	for (; what[i] >= 0 ; i++)
	{
		unsigned int idx = what[i] ;
		(*opts_tree_table[idx].func)(fields[opts_tree_table[idx].id],treename) ;
	}
	
}

static void info_parse_options(char const *str,int *what)
{
	size_t pos = 0 ;
	stralloc sa = STRALLOC_ZERO ;
	
	if (!sastr_clean_string_wdelim(&sa,str,DELIM)) log_dieu(LOG_EXIT_SYS,"parse options") ;
	unsigned int n = sastr_len(&sa), nopts = 0 , old ;
	checkopts(n) ;
	info_opts_map_t const *t ;
	
	for (;pos < sa.len; pos += strlen(sa.s + pos) + 1)
	{
		char *o = sa.s + pos ;
		t = opts_tree_table ;
		old = nopts ;
		for (; t->str; t++)
		{			
			if (obstr_equal(o,t->str))
			{
				/*if (!t->id){ what[0] = t->id ; old++ ; }
				else*/ what[nopts++] = t->id ;
			}
		}
		if (old == nopts) log_die(LOG_EXIT_SYS,"invalid option: ",o) ;
	}
	
	stralloc_free(&sa) ;
}

int main(int argc, char const *const *argv, char const *const *envp)
{
	unsigned int legacy = 1 ;
	
	size_t pos, newlen, livelen ;
	int what[MAXOPTS] = { 0 } ;
	
	log_color = &log_color_disable ;
	
	char const *treename = 0 ;
	
	for (int i = 0 ; i < MAXOPTS ; i++)
		what[i] = -1 ;
		
	
	char buf[MAXOPTS][INFO_FIELD_MAXLEN] = {
		"Name",
		"Initialized",
		"Enabled",
		"Starts after",
		"Current",
		"Allowed",
		"Symlinks",
		"Contents" } ;
	
	
	stralloc satree = STRALLOC_ZERO ;
	
	PROG = "66-intree" ;
	{
		subgetopt_t l = SUBGETOPT_ZERO ;

		for (;;)
		{
			int opt = getopt_args(argc,argv, ">hzv:no:grd:l:c", &l) ;
			if (opt == -1) break ;
			if (opt == -2) log_die(LOG_EXIT_USER,"options must be set first") ;
			switch (opt)
			{
				case 'h' : 	info_help(); return 0 ;
				case 'v' :  if (!uint0_scan(l.arg, &VERBOSITY)) log_usage(USAGE) ; break ;
				case 'z' :	log_color = !isatty(1) ? &log_color_disable : &log_color_enable ; break ;
				case 'n' :	NOFIELD = 0 ; break ;
				case 'o' : 	legacy = 0 ; info_parse_options(l.arg,what) ; break ;
				case 'g' :	GRAPH = 1 ; break ;
				case 'r' : 	REVERSE = 1 ; break ;
				case 'd' : 	if (!uint0_scan(l.arg, &MAXDEPTH)) log_usage(USAGE) ; break ;
				case 'l' : 	if (!stralloc_cats(&live,l.arg)) log_usage(USAGE) ;
							if (!stralloc_0(&live)) log_usage(USAGE) ;
							break ;
				case 'c' :	log_die(LOG_EXIT_SYS,"deprecated option -- please use -z instead") ;
				default : 	log_usage(USAGE) ; 
			}
		}
		argc -= l.ind ; argv += l.ind ;
	}
	
	if (argv[0]) treename = argv[0] ;
	
	if (legacy)
	{
		unsigned int i = 0 ;
		for (; i < MAXOPTS - 1 ; i++)
			what[i] = i ;
			
		what[i] = -1 ;
	}
	
	OWNER = getuid() ;
	size_t ownerlen = uid_fmt(OWNERSTR,OWNER) ;
	OWNERSTR[ownerlen] = 0 ;
	
	info_field_align(buf,fields,field_suffix,MAXOPTS) ;
		
	setlocale(LC_ALL, "");
	
	if(!str_diff(nl_langinfo(CODESET), "UTF-8")) {
		STYLE = &graph_utf8;
	}
	
	if (!set_ownersysdir(&base,OWNER)) log_dieusys(LOG_EXIT_SYS, "set owner directory") ;
	if (!stralloc_copy(&src,&base) ||
	!stralloc_cats(&src,SS_SYSTEM) ||
	!stralloc_0(&src)) log_die_nomem("stralloc") ;
	src.len-- ;
	
	if (!scan_mode(src.s,S_IFDIR))
	{
		log_info("no tree exist yet") ;
		goto freed ;
	}
	
	if (!stralloc_cats(&src,"/")) log_die_nomem("stralloc") ;
	
	newlen = src.len ;
	livelen = live.len ;

	if (treename)
	{
		
		if (!auto_stra(&src,treename)) log_die_nomem("stralloc") ;
		if (!scan_mode(src.s,S_IFDIR)) log_dieusys(LOG_EXIT_SYS,"find tree: ", src.s) ;
		src.len = newlen ;
		if (!stralloc_0(&src)) log_die_nomem("stralloc") ;
		src.len-- ;
		info_display_all(treename,what) ;		
	}
	else
	{
		if (!stralloc_0(&src)) log_die_nomem("stralloc") ;
	    if (!sastr_dir_get(&satree, src.s,SS_BACKUP + 1, S_IFDIR)) log_dieusys(LOG_EXIT_SYS,"get list of tree at: ",src.s) ;
		if (satree.len)
		{
			if (!info_cmpnsort(&satree)) log_dieu(LOG_EXIT_SYS,"sort list of tree") ;
			for(pos = 0 ; pos < satree.len ; pos += strlen(satree.s + pos) +1 )
			{
				src.len = newlen ;
				live.len = livelen ;
				char *name = satree.s + pos ;
				info_display_all(name,what) ;
				if (buffer_puts(buffer_1,"\n") == -1) 
					log_dieusys(LOG_EXIT_SYS,"write to stdout") ;	
			}
		}
		else 
		{
			log_info("no tree exist yet") ;
			goto freed ;
		}
	}

	if (buffer_flush(buffer_1) == -1) 
		log_dieusys(LOG_EXIT_SYS, "write to stdout") ;
	
		
	freed:
	stralloc_free(&base) ;
	stralloc_free(&live) ;
	stralloc_free(&src) ;
	stralloc_free(&satree) ;
		
	return 0 ;
}
